### DISCUSSION TITLE
Oracle Database & Application-Consistent Backups

### DISCUSSION THREAD

**Discussion by csl04r** [longhorn/longhorn#4101] [Q&A]
# Does Longhorn Guarantee "Write Ordering"?
This is the question my Oracle DBA asked me. I don't know the answer. I don't even understand the question fully. 

I asked myself instead, would a longhorn snapshot taken without reference to the state of the database be useful? Is it different than if the power were suddenly cut from a machine running Oracle database with a traditional direct-attached volume? I suspect that longhorn volumes _are_ different, owing to the replication across the network. But I'm just an app developer, I don't really know. I suspect that if the node hosting the db or volume crashed, the head of the replicas might not be useful. I'd love for someone to show me how I'm worried about nothing.

# My Goal
I am trying out longhorn to supply a 3x replicated PV for an oracle database pod. 
My goal is to make frequent (every 5 minutes) *snapshots*, and daily *backups* of this database's data.

# Working with the Database
Not knowing the answers to those questions, or a good way to test and get confidence, I instead found that Oracle supports user-managed backups. Provided you put the database into backup mode, you can perform an OS- or Volume-level backup, then take the database out of backup mode, e.g.

```
sqlplus <<< 'alter database begin backup;'
tar zcvf /backupstore/orabak.tar.gz ./oradata
sqlplus <<< 'alter database end backup;'
```

Is there some longhorn can bracket the creation of a snapshot or backup with pre- and post- commands in another pod?

# What I Tried

I can trigger a snapshot with HTTP if I have the volume name, for instance with `curl` running in the database pod. For the sake of brevity it ends up being something like this (ignoring errors):

```
# prepare db for volume snapshot
sqlplus backupuser/backuppass@db as sysbackup <<< 'alter database begin backup;'
# ask longhorn to create volume snapshot
curl \
        --header accept:application/json \
        --include \
        --data '{}' \
        --location \
        --fail \
        "http://longhorn-frontend.longhorn-system.svc/v1/volumes/pvc-blahblahblabhblabh?action=snapshotCreate"
# tell db to resume normal operation
sqlplus backupuser/backuppass@db as sysbackup <<< 'alter database end backup;'
```

# Asynchronicity
I noticed when I tried to do the same for creating a _backup_, the `curl` command returns right away. The backup happens asynchronously and takes an amount of time dependent on how large the volume is. 

I need to wait for the backup to finish completely before telling the db to `end backup`.
I didn't notice this issue with snapshots because they are made quite quicly, I could probably `sleep 10` or poll longhorn for the state of the snapshot/backup before telling DB to resume normal operations. This is probably easier said than done in a truly robust way.

# Feature Request
I see plenty of advice on how to automate backups using `RecurringJob` or scheduling them via the UI. I see no advice or acknowledgement that these actions need to be coordinated with workloads, esp. database workloads. 

So unless I'm missing something I'd like to request some way to do this-- to get longhorn to manage the lifecycle of snapshot and backup creation, allowing some kind of arbitrary hook to be executed before and after.

I'm open to hear other suggestions on how to do this.

# Other ideas
I could use longhorn for snapshots, and use oracle facilities `rman` for full backups. I'd rather not. Life will be simpler I think, if all backup/restore can be done via longhorn.

**Comment by derekbit** [comment]
IIUC, the main question you're asking is if Longhorn supports the application-consistent backup and if Longhorn has the hooks similar to the velero hooks that make the application's backup job work with Longhorn. Is it correct?

If yes, we have a [ticket](https://github.com/longhorn/longhorn/issues/2128) for this feature.
